### DISCUSSION TITLE
Batch label for tar_stan_mcmc_rep_summary

### DISCUSSION THREAD

**Discussion by mike-lawrence** [wlandau/stantargets#25] [Help] [answered]
When I use a `tar_stan_mcmc_rep_summary` target for SBC, I see there's a `.rep` column but I'm not seeing a `.batch` column; is there a way to get a column indicating which batch a given chunk of the summary comes from?

**Comment by wlandau** [comment, marked as answer]
There is no `.batch` column or similar. Would that be useful? The batch just indicates how the computation was parallelized and has no intrinsic meaning as far as the results are concerned.

**Reply by mike-lawrence** [reply]
Ah, my bad. I had been misunderstanding the relation between batch, rep and chain. I see now that the .rep label is probably all I'd ever need.

**Reply by mike-lawrence** [reply]
Actually, on further consideration, since each batch is associated with a single simulated dataset, and reps re-use this dataset, it could be useful to have a batch identifier after all. The use case for using reps at all (assuming using `chains>=1`) would be to explore the influence of different warmup strategies, in which case the use case for a batch identifier would be to account for possible weirdness of a specific dataset that will affect all associated reps.

**Reply by wlandau** [reply]
> since each batch is associated with a single simulated dataset...

Each rep within batch gets its own dataset, and batches have multiple datasets.

https://github.com/wlandau/stantargets/blob/df816f34ce45dcf61a46295ad37db522520b6eb2/R/tar_stan_mcmc_rep.R#L122-L130

The purpose of [batching](https://books.ropensci.org/targets/dynamic.html#batching) is to [define good targets](https://books.ropensci.org/targets/practices.html#how-to-define-good-targets) that balance the overhead of scale vs vulnerability to invalidation. Otherwise, batches have no meaning.

``` r
library(targets)
library(stantargets)
tar_stan_example_file("a.stan")
  
tar_script({
library(stantargets)
tar_stan_mcmc_rep_summary(
  your_model,
  stan_files = "a.stan", 
  data = tar_stan_example_data(),
  batches = 1,
  reps = 2,
  stdout = R.utils::nullfile(),
  stderr = R.utils::nullfile()
)
})

tar_make(reporter = "silent")

str(tar_read(your_model_data, branches = 1))
#> List of 1
#>  $ :List of 2
#>   ..$ :List of 5
#>   .. ..$ n         : int 10
#>   .. ..$ x         : num [1:10] -1 -0.778 -0.556 -0.333 -0.111 ...
#>   .. ..$ y         : num [1:10] -0.145 1.156 0.675 0.46 0.639 ...
#>   .. ..$ true_beta : num -0.0748
#>   .. ..$ .join_data:List of 2
#>   .. .. ..$ beta : num -0.0748
#>   .. .. ..$ y_rep: num [1:10] -0.145 1.156 0.675 0.46 0.639 ...
#>   ..$ :List of 5
#>   .. ..$ n         : int 10
#>   .. ..$ x         : num [1:10] -1 -0.778 -0.556 -0.333 -0.111 ...
#>   .. ..$ y         : num [1:10] -1.32 -1.55 -2.06 -1.36 1.18 ...
#>   .. ..$ true_beta : num 0.713
#>   .. ..$ .join_data:List of 2
#>   .. .. ..$ beta : num 0.713
#>   .. .. ..$ y_rep: num [1:10] -1.32 -1.55 -2.06 -1.36 1.18 ...
```

<sup>Created on 2021-03-23 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Reply by mike-lawrence** [reply]
Ah! I clearly *still* had the wrong understanding, thanks for clarifying.

**Reply by mike-lawrence** [reply]
I think the help docs were what led me astray; under the entry for the batches argument it says "Number of batches. Each batch is a branch target that generates a dataset and runs the model reps times."

**Reply by wlandau** [reply]
Oh, right, that's definitely grammatically ambiguous. Now fixed.
